pageserver/src/storage_sync.rs

```diff
@@ -129,1330 +114,587 @@
+        // Launch the tasks immediately, if possible
+        self.launch_queued_tasks(upload_queue);
+        Ok(())
     }
 
-    Ok(())
-}
+    ///
+    /// Wait for all previously scheduled uploads/deletions to complete
+    ///
+    pub async fn wait_completion(self: &Arc<Self>) -> Result<()> {
+        let (sender, mut receiver) = tokio::sync::watch::channel(());
+        let barrier_op = UploadOp::Barrier(sender);
 
-async fn delete_timeline_data(
-    conf: &'static PageServerConf,
-    (storage, index, sync_queue): (&GenericRemoteStorage, &RemoteIndex, &SyncQueue),
-    sync_id: TenantTimelineId,
-    mut new_delete_data: SyncData<LayersDeletion>,
-    sync_start: Instant,
-) {
-    static TASK_NAME: &str = "delete";
-
-    let timeline_delete = &mut new_delete_data.data;
-
-    if !timeline_delete.deletion_registered {
-        if let Err(e) = update_remote_data(
-            conf,
-            storage,
-            index,
-            sync_id,
-            RemoteDataUpdate::Delete(&timeline_delete.layers_to_delete),
-        )
-        .await
         {
-            error!("Failed to update remote timeline {sync_id}: {e:?}");
-            new_delete_data.retries += 1;
-            sync_queue.push(sync_id, SyncTask::Delete(new_delete_data));
-            register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
-            return;
+            let mut upload_queue = self.upload_queue.lock().unwrap();
+            ensure!(
+                upload_queue.latest_metadata.is_some(),
+                "upload queue not initialized"
+            );
+
+            upload_queue.queued_operations.push_back(barrier_op);
+
+            // Launch the task immediately, if possible
+            self.launch_queued_tasks(upload_queue);
         }
+
+        receiver.changed().await?;
+        Ok(())
     }
-    timeline_delete.deletion_registered = true;
 
-    let sync_status = delete_timeline_layers(storage, sync_queue, sync_id, new_delete_data).await;
-    register_sync_status(sync_id, sync_start, TASK_NAME, Some(sync_status));
-}
+    ///
+    /// Pick next tasks from the queue, and start as many of them as possible without violating
+    /// the ordering constraints.
+    ///
+    /// The caller needs to already hold the `upload_queue` lock.
+    fn launch_queued_tasks(self: &Arc<Self>, mut upload_queue: MutexGuard<UploadQueue>) {
+        while let Some(next_op) = upload_queue.queued_operations.front() {
+            // Can we run this task now?
+            let can_run_now = match next_op {
+                UploadOp::UploadLayer(_) => {
+                    // Can always be scheduled.
+                    true
+                }
+                UploadOp::UploadMetadata(_, _) => {
+                    // These can only be performed after all the preceding operations
```

Comment:
I'm not entirely sure this is ok: the deletion operation requires the metadata to be updated first.
In `schedule_layer_file_deletion` we do 
```rust
upload_queue
            .queued_operations
            .push_back(UploadOp::UploadMetadata(index_part, disk_consistent_lsn));

        // schedule the actual deletions
        for path in paths {
            upload_queue
                .queued_operations
                .push_back(UploadOp::Delete(PathBuf::from(path)));
            info!("scheduled layer file deletion {}", path.display());
        }
```

with this logic, it loos like, we're actually waiting for the deletions to happen first, despite the scheduling?

Reply:
I don't see the problem, can you elaborate?

The current logic is that before performing index-file upload, we wait for all operations that were scheduled earlier to finish. And `schedule_layer_file_deletion` schedules the index-file upload before the deletions. Seems correct to me.

Reply:
> Seems correct to me.

Indeed, I was wrong.